theano/tensor/nnet/abstract_conv.py

```diff
@@ -177,6 +177,187 @@ def conv2d(input,
     return conv_op(input, filters)
 
 
+def deconv(input,
+           filters,
+           input_shape=None,
+           output_shape=None,
+           filter_shape=None,
+           border_mode='valid',
+           subsample=(1, 1),
+           filter_flip=True):
+    """This function will build the symbolic graph for deconvoloving a
+    mini-batch of a stack of 2D inputs with a set of 2D filters (using
+    gradient of convolution), such that it upsamples the input to the
+    desired output resolution.
+
+    :type input: symbolic 4D tensor to be upsampled
+    :param input: mini-batch of feature map stacks, of shape
+        (batch size, input channels, input rows, input columns).
+        See the optional parameter ``input_shape``.
+
+    :type filters: symbolic 4D tensor
+    :param filters: set of filters used in CNN layer of shape
+        (output channels, input channels, filter rows, filter columns).
+        See the optional parameter ``filter_shape``.
+
+    :type input_shape: None, tuple/list of len 4 of int or Constant variable
+        indicating the shape to be upsampled or whose gradient should be taken.
+    :param input_shape: The shape of the input parameter.
+        Optional, possibly used to choose an optimal implementation.
+        You can give ``None`` for any element of the list to specify that this
+        element is not known at compile time.
+
+    :type output_shape: tuple/list of len 2 of int or Constant variable
+        indicating the row and column size of the output (upsampled) features.
+    :param output_shape: The shape of the input parameter.
+        Not Optional, since given the input_shape and the subsample values,
+        multiple output_shape may be plausible.
+
+    :type filter_shape: None, tuple/list of len 4 of int or Constant variable
+    :param filter_shape: The shape of the filters parameter.
+        Optional, possibly used to choose an optimal implementation.
+        You can give ``None`` for any element of the list to specify that this
+        element is not known at compile time.
+
+    :type border_mode: str, int or tuple of two int
+    :param border_mode: Either of the following:
+        * ``'valid'``: apply filter wherever it completely overlaps with the
+          input. Generates output of shape: input shape - filter shape + 1
+        * ``'full'``: apply filter wherever it partly overlaps with the input.
+          Generates output of shape: input shape + filter shape - 1
+        * ``'half'``: pad input with a symmetric border of ``filter rows // 2``
```

Comment:
`same` is tricky, because it may mean uneven padding on both sides, which is not implemented in some back-ends (if not all), or taking a subtensor (which makes the tensor non-contiguous).
This is why we have `half`, which is equivalent to `same` when `same` is achievable with symmetric padding.